### PR TITLE
gm-editor add autoupdating while game unpaused and memory staleness c…

### PR DESCRIPTION
**Pull request by sdegrace**
…hecking

Fixes https://github.com/DFHack/dfhack/issues/2801

### CODE REVIEW

File: gui/gm-editor.lua [hunk 1]
```diff
@@ -453,6 +503,13 @@ function GmEditorUi:onInput(keys)
     elseif keys[keybindings.help.key] then
         self.subviews.pages:setSelected(2)
         return true
+	elseif keys[keybindings.autoupdate.key] then
+		if not self.autoupdate then
+			self.autoupdateLabel:setText({{gap=1,text="Auto-Update running... ",key=keybindings.autoupdate.key,key_sep = '()'}})
+		else
+			self.autoupdateLabel:setText({{gap=1,text="Auto-Update stopped...",key=keybindings.autoupdate.key,key_sep = '()'}})
+		end
+		self.autoupdate = not self.autoupdate
```
Comment (myk002) [inline gui/gm-editor.lua:512]: This can be removed if you use a widgets.ToggleHotkeyLabel`

File: gui/gm-editor.lua [hunk 2]
```diff
@@ -545,6 +604,12 @@ function GmEditorUi:postUpdateLayout()
     config:write(self.frame)
 end
 
+function GmEditorUi:onRenderBody()
+    if self.next_refresh_ms <= dfhack.getTickCount() and self.autoupdate then
```
Comment (myk002) [inline gui/gm-editor.lua:608]:
```suggestion
    if self.subviews.lbl_autoupdate:getOptionValue() and self.next_refresh_ms <= dfhack.getTickCount() then
```

Comment (myk002) [reply gui/gm-editor.lua:608]: to avoid the call through the lua-c++ layer if it's not necessary

File: gui/gm-editor.lua [hunk 3]
```diff
@@ -135,16 +140,56 @@ function GmEditorUi:init(args)
...
         on_submit2=self:callback("editSelectedRaw"),
         text_pen=COLOR_GREY, cursor_pen=COLOR_YELLOW}
+	self.autoupdateLabel = widgets.Label{text={{gap=1,text="Auto-Update stopped...",key=keybindings.autoupdate.key,key_sep = '()'}}, view_id = 'lbl_autoupdate',frame = {l=0,t=0,yalign=0}}
     local mainPage=widgets.Panel{
         subviews={
             mainList,
             widgets.Label{text={{text="<no item>",id="name"},{gap=1,text="Help",key=keybindings.help.key,key_sep = '()'}}, view_id = 'lbl_current_item',frame = {l=1,t=1,yalign=0}},
-            widgets.EditField{frame={l=1,t=2,h=1},label_text="Search",key=keybindings.start_filter.key,key_sep='(): ',on_change=self:callback('text_input'),view_id="filter_input"}}
+            widgets.EditField{frame={l=1,t=2,h=1},label_text="Search",key=keybindings.start_filter.key,key_sep='(): ',on_change=self:callback('text_input'),view_id="filter_input"},
+			self.autoupdateLabel}
```
Comment (myk002) [inline gui/gm-editor.lua:149]:
```suggestion
            widgets.ToggleHotkeyLabel{view_id='lbl_autoupdate', frame={r=1, t=1}, auto_width=1, label='Auto-refresh live data', key=keybindings.autoupdate.key, key_sep='()', initial_option=false},
        }
```

File: gui/gm-editor.lua [hunk 4]
```diff
@@ -135,16 +140,56 @@ function GmEditorUi:init(args)
     local mainList=widgets.List{view_id="list_main",choices={},frame = {l=1,t=3,yalign=0},on_submit=self:callback("editSelected"),
         on_submit2=self:callback("editSelectedRaw"),
         text_pen=COLOR_GREY, cursor_pen=COLOR_YELLOW}
+	self.autoupdateLabel = widgets.Label{text={{gap=1,text="Auto-Update stopped...",key=keybindings.autoupdate.key,key_sep = '()'}}, view_id = 'lbl_autoupdate',frame = {l=0,t=0,yalign=0}}
```
Comment (myk002) [inline gui/gm-editor.lua:143]:
```suggestion
```

File: gui/gm-editor.lua [hunk 5]
```diff
@@ -122,6 +126,7 @@ function GmEditorUi:init(args)
     self.stack={}
     self.item_count=0
     self.keys={}
+	self.autoupdate = false
```
Comment (myk002) [inline gui/gm-editor.lua:129]:
```suggestion
```

File: gui/gm-editor.lua [hunk 6]
```diff
@@ -135,16 +140,56 @@ function GmEditorUi:init(args)
...
+			end
+		)
+		if not status then
+			failure = true
+			last_good_level = i - 1
+			break
+		end
+
+
+		if not next_in_stack == next_by_ref then
```
Comment (myk002) [inline gui/gm-editor.lua:181]: where is next_in_stack set?

File: gui/gm-editor.lua [hunk 7]
```diff
@@ -135,16 +140,56 @@ function GmEditorUi:init(args)
...
+	for i, level in pairs(self.stack) do
+		local obj=level.target
+
+		local keys = level.keys
+		local selection = level.selected
+		local sel_key = keys[selection]
+		local next_by_ref
+		local status, _ = pcall(
+			function()
+				next_by_ref = obj[sel_key]
```
Comment (myk002) [inline gui/gm-editor.lua:169]: in addition to checking whether the key exists, can we also check to make sure it points to the same place it used to? for example, will this detect when the element at n-1 was deleted and everything just moved down a slot and the struct that we're looking at is no longer at index n?

**Review by myk002 — COMMENTED** [review]
suggested a different widget to cut down on manual state tracking

could you fix up the formatting so the lint and ci_precommit checks pass?

File: gui/gm-editor.lua (hunk 7 above)
Comment (sdegrace) [reply gui/gm-editor.lua:169]: This was meant to be addressed by next_in_stack - see below

**Review by myk002 — COMMENTED** [review]
I need to playtest this a bit more, once DFHack is better aligned with DF 50.07 structures but overall looks good!

**Comment by myk002** [comment]
One thing I've noticed is that the selection gets reset to the first element on every refresh. Could you ensure that the selected element stays constant as long as the stack doesn't need adjustment?

File: gui/gm-editor.lua [hunk 8]
```diff
@@ -9,6 +9,9 @@ local utils = require 'utils'
 
 config = config or json.open('dfhack-config/gm-editor.json')
 
+local REFRESH_MS = 100
```
[the remark this answers is not in the document]
Comment (myk002) [reply gui/gm-editor.lua:12]: edit: 100 ms is actually quite nice to track things that move quickly. I'd say keep this as it is.

File: gui/gm-editor.lua [hunk 9]
```diff
@@ -545,11 +589,11 @@ function GmEditorUi:postUpdateLayout()
...
-GmScreen = defclass(GmScreen, gui.ZScreen)
-GmScreen.ATTRS {
-    focus_path='gm-editor',
-}
-
+function GmEditorUi:onRenderBody()
+    if self.subviews.lbl_autoupdate:getOptionValue() and self.next_refresh_ms <= dfhack.getTickCount()  then
+        self:updateTarget()
+    end
+end
```
Comment (myk002) [inline gui/gm-editor.lua:596]:
```suggestion
end

GmScreen = defclass(GmScreen, gui.ZScreen)
GmScreen.ATTRS {
    focus_path='gm-editor',
}

```

**Comment by myk002** [comment]
ping on this

**Comment by myk002** [comment]
any progress on this PR? it's something that would be wonderful to have in the DFHack library

**Review by myk002 — APPROVED** [review]

**Pull request merged**